mitodl/mitxpro
# Fix tracking of course run selections when completing orders

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes https://trello.com/c/2revzJqS/110-look-into-cause-of-completed-program-orders-with-no-course-run-selections

#### What's this PR do?
Fixes bug where users who purchased a program would sometimes end up with no course run selections. Also makes a few refactors to the checkout code

#### How should this be manually tested?
- Complete an order for a program. You should end up with `CourseRunEnrollment`s for all the runs you selected, and you should also see `LineRunSelections` that correspond to the runs you selected (`/admin/ecommerce/linerunselection/`)
- Also complete an order for a course and make sure those same artifacts are created

#### Any background context you want to provide?
- The cause of this bug was that we were depending on a `Basket` and associated `CourseRunSelection`s when an order was being completed to determine what course runs were selected and needed enrollments. The problem with that is that a basket can be updated simply by visiting and interacting with the checkout page. If an order was submitted, then a user visited the checkout page before the order was actually completed (which is easily possible if we're waiting for Cybersource), the basket would be updated and the wrong enrollments could be created (or none at all)
- I made a few refactors to the checkout logic while I was in there. We were repeating assumptions about the basket data in several places and going back to the database several times over for objects that were already found. There were performance concerns and high potential for bugs, so I went ahead and refactored that section and associated tests
- I edited a bunch of validation message that read like developer debugging messages but are shown to users. For example, a lot of messages made references to "runs". Our users don't know about the concept of course runs vs. courses, nor do they need to.
